File: Algorithms/PriorityQueue/PriorityQueue.py
class PriorityQueue:
    def __init__(self):
        self.__queue = []
        self.__index = 0
    
    def push(self, item, priority):
        from heapq import heappush
        heappush(self.__queue, (-priority, self.__index, item))
        self.__index += 1

    def pop(self):
        from heapq import heappop
        return heappop(self.__queue)[-1]


def main():
    commands_num = int(input())
    priority_queue = PriorityQueue()
    extracted_items = []

    for _ in range(commands_num):
        commamd = input()
        if commamd.startswith("Insert"):
            _, item = commamd.split()
            priority = item = int(item)
            priority_queue.push(item, priority)
        else:
            extracted_items.append(priority_queue.pop())

    for extracted_item in extracted_items:
        print(extracted_item)


if __name__ == '__main__':
    main()

File: Algorithms/PriorityQueue/PriorityQueue.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>

template <typename Item>
class PriorityQueue {
    private:
        std::priority_queue<Item> queue;
        std::vector<Item> extractedCache;

    public:
        void push(int priority) {
            queue.push(priority);
        }

        void pop() {
            extractedCache.push_back(queue.top());
            queue.pop();
        }

        std::pair<std::string, int> parseCommand(std::string line, std::string delimiter) {
            std::string command;
            int priority;

            auto start = 0U;
            auto end = line.find(delimiter);

            while (end != std::string::npos) {
                command = line.substr(start, end - start);
                start = end + delimiter.length();
                end = line.find(delimiter, start);
            }

            if (line != "ExtractMax") {
                priority = std::stoi(line.substr(start, end));
            } else {
                priority = 0;
            }
            
            return {command, priority};
        }

        void printExtractedCache() {
            for (auto item : extractedCache) {
                std::cout << item << std::endl;
            }
        }
};

int main() {
    int commandsNumber;
    std::cin >> commandsNumber;
    PriorityQueue<int> priorityQueue;

    for (int i = 0; i <= commandsNumber; i++) {
        std::string line;
        std::getline(std::cin, line);

        std::string command;
        int priority;
        if (!line.empty()) {
            std::tie(command, priority) = priorityQueue.parseCommand(line, " ");
            if (command == "Insert") {
                priorityQueue.push(priority);
            } else {
                priorityQueue.pop();
            }
        }
    }

    priorityQueue.printExtractedCache();
    
    return 0;
}